pkg/build/builder/dockerutil.go

```diff
@@ -62,15 +73,59 @@ func pushImage(client DockerClient, name string, authConfig docker.AuthConfigura
 	if glog.V(5) {
 		opts.OutputStream = os.Stderr
 	}
-	var err error
+
 	var retriableError = false
 
+	eventWatch, err := eventClient.Watch(
+		kapi.ListOptions{
+			FieldSelector: fields.ParseSelectorOrDie(
+				strings.Join([]string{
+					"involvedObject.kind==" + "ImageStream",
+					"involvedObject.namespace==" + ref.Namespace,
+					"involvedObject.name==" + ref.Name},
+					",")),
+		},
+	)
+	if err != nil {
+		glog.Errorf("Failed to watch for events: %v", err)
+	} else {
+		glog.V(5).Infof("Listening for events on image stream %s", ref.RepositoryName())
+		defer eventWatch.Stop()
+	}
+
 	for retries := 0; retries <= DefaultPushRetryCount; retries++ {
 		err = client.PushImage(opts, authConfig)
 		if err == nil {
 			return nil
 		}
 
+		waitEnd := time.Now().Add(time.Second * 2)
+
+		if eventWatch != nil && strings.Contains(err.Error(), "denied") {
```

Comment:
This is not a kAPI error. The error message this is trying to match is https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/docker/distribution/registry/api/errcode/register.go#L52.

I'd like to avoid putting here whole error message because it may get changed during a next docker/distribution rebase.